### ISSUE TITLE
Test failing depending o Kernel and OS configurations

### ISSUE THREAD

**Issue by nawarian** [reactphp/event-loop#146]
Hi, all!

I was running the tests at my local environment because I wanted to improve somethings and I like to start by making sure everything is running properly ^^
Thing is that the test `\React\Tests\EventLoop\AbstractLoopTest::testRemoveReadAndWriteStreamFromLoopOnceResourceClosesOnEndOfFileEndsLoop` is failing for all implementations with the following message:

```
fwrite(): send of 8192 bytes failed with errno=35 Resource temporarily unavailable
```

This happens at [line 231](https://github.com/reactphp/event-loop/blob/master/tests/AbstractLoopTest.php#L231), which runs this: `fwrite($other, str_repeat('.', 60000));`

I did some basic research and looks like Mac OS can have very low values for buffer sizes, but changing those values via `sysctl` didn't take much effect.

According to [this errors list](http://krypted.com/lists/comprehensive-list-of-mac-os-x-error-codes/) errno=35 Resource temporarily unavailable means `kPOSIXErrorEAGAIN` and according to [here](http://www.gnu.org/software/libc/manual/html_node/Error-Codes.html) the `EAGAIN` error says that possibly it would work in further attempts (not desirable for unit testing) and can be caused because `a)` a resource could be doing a blocking operation and was set as non-blocking before, which is plausible because for both streams we call `stream_set_blocking` with `false` or `b)` due to a resource shortage, which for me would make sense with the `sysctl` stuff. Documentation says that maybe delaying certain operations would help.

I suppose this kind of error would be super hard to reproduce, but I have something else that may help:

```
nawarian@phoenix event-loop (master) $ sysctl -a | egrep "space|maxsockbuf"
kern.ipc.maxsockbuf: 8388608
net.local.stream.recvspace: 8192
net.local.stream.sendspace: 8192
net.local.dgram.recvspace: 4096
net.inet.tcp.sendspace: 131072
net.inet.tcp.recvspace: 131072
net.inet.udp.recvspace: 196724
net.inet.raw.recvspace: 8192
net.stats.sendspace: 2048
net.stats.recvspace: 8192
```

It might be possible to reproduce such error if you run something like `sysctl -w variable.name=new-value` using those keys and values. But again: I'm not sure.

Could also be useful:

```
nawarian@phoenix event-loop (master) $ sw_vers
ProductName:	Mac OS X
ProductVersion:	10.12.6
BuildVersion:	16G1114
```

Oh yes, almost forgot: lowering the string length at `fwrite($other, str_repeat('.', 60000));` solve the issue.

**Comment by nawarian** [comment]
Further tests...

Changing the write line to `fwrite($other, str_repeat('.', 8192));` works perfectly. Changing to `fwrite($other, str_repeat('.', 8193));` will result at this message:

```
fwrite(): send of 1 bytes failed with errno=35 Resource temporarily unavailable
```

So it is chunking by 8192 bites per time. But seems like the second chunk (1 remaining byte) cannot be sent (maybe the resource is not available for writing at that moment).

**Comment by kelunik** [comment]
You correctly observed that the macOS write buffer defaults to 8192 bytes. The quick fix here would be either reducing the number of bytes to 8192 or setting the `$other` stream to blocking, which should be fine for that test. However, that that might break depending on the event loop implementation, because it reads only `1024` bytes while the default chunk size for PHP is `8192`. If PHP reads more than `1024` bytes and no further data is incoming (which is not the case here, we have enough data), that might not trigger the readability watcher, as PHP has that data already in its buffer instead of it being in the kernel's buffer.

**Comment by nawarian** [comment]
I'm sure it will solve the test, but not sure if it will solve the problem itself.

By attempting to write `60000` bytes to `$other` even if you try to read `$stream` by `1024` chunks it will have the blocking issue because the Unix sock itself is not ready. Notice that if you set `$other` to blocking the same test will take ages to run.
I think MacOS implementation simply won't allow this kind of thing.

In any case, I'll open a pull request to fix this test but think would be a good idea to think about another implementation for it (which I'd love to propose but I don't have much experience with socks and stuff).

**Comment by kelunik** [comment]
The solution is basically https://github.com/reactphp/stream or a simple write watcher and a buffer.

**Comment by clue** [comment]
@nawarian Thank you for reporting, your suggested change makes perfect sense to me :+1: 

Closed via #147
